### PR TITLE
recover from client-caused panics

### PR DESCRIPTION
**Pull request by slingamn**
I accidentally wrote an edge case with a nil dereference over on #155. Then I fixed it, but it seems like we can mostly just work around them using the `recover()` function.

To test this change, I reverted 2791476f16634ae664fbd86a946de09fe1f7fa9b and triggered the underlying crash. The following output was logged (the first line via the logging mechanism, and the rest to stderr directly): https://gist.github.com/slingamn/661b326b5b0cafca36fcb8cb7e4dde17

Other clients, and the overall server state, were not disrupted.

Any concerns about logging private data? It's vaguely possible that something private could be in the string value of the panic itself, but the stack trace seems fine to me.

### CODE REVIEW

**Comment by slingamn** [comment]
I can also see a case for this being configurable.

File: irc/client.go [hunk 1]
```diff
@@ -184,6 +184,15 @@ func (client *Client) run() {
 	var line string
 	var msg ircmsg.IrcMessage
 
+	defer func() {
+		if r := recover(); r != nil {
+			client.server.logger.Error("internal", fmt.Sprintf("Client caused panic, disconnecting: %v", r))
+			debug.PrintStack()
```
Comment (DanielOaks) [inline irc/client.go]: Hmm... I like this, but I think it makes more sense to log this whole string as an error -- makes digging into submitted debug logs a whole lot easier later. Could we do something along the lines of this?
```golang
client.server.logger.Error("internal", fmt.Sprintf("Client caused panic, disconnecting: %v\n%v", r, debug.Stack()))
```

**Comment by DanielOaks** [comment]
Awesome. There is a concern for logging private data, but... I think that it's more important to ensure we're able to fix bugs. The panic shouldn't _really_ contain sensitive data most of the time, and I think it's worth the trade-off (particularly since we're still an experimental server that does have some dodgy bugs we're shaking out).

**Comment by slingamn** [comment]
Another possible reason to make this configurable: for the benefit of people who would rather have the server crash (and restart, depending on their configuration) than have it left in a potentially inconsistent state. (The panic/recover mechanism unwinds the stack and executes all deferred statements as it goes, but it's still possible that after the `recover()`, some kind of invariant is violated or one of the locks is stuck in the locked state.)

File: irc/client.go [hunk 2]
```diff
@@ -185,9 +185,11 @@ func (client *Client) run() {
 	var msg ircmsg.IrcMessage
 
 	defer func() {
-		if r := recover(); r != nil {
-			client.server.logger.Error("internal",
-				fmt.Sprintf("Client caused panic, disconnecting: %v\n%s", r, debug.Stack()))
+		if client.server.RecoverFromErrors() {
+			if r := recover(); r != nil {
+				client.server.logger.Error("internal",
+					fmt.Sprintf("Client caused panic, disconnecting: %v\n%s", r, debug.Stack()))
```
Comment (DanielOaks) [inline irc/client.go]: Hmm... could we do something along the lines of:

1. If RecoverFromErrors is true, recover and log the error.
2. If RecoverFromErrors is false, log the error and then let it kill the server.

I'd sorta like to always know what caused the server to fail, even if we choose to kill the server anyways right after. Maybe if we can't re-raise the error, we could just log the stack as above, then shutdown the server manually?

**Comment by slingamn** [comment]
Re-panicking seems to work OK: https://gist.github.com/slingamn/bc4886682c8834925c4e3343b89a031b
